url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/protocols/specifications/PoulidorProtocol.java

```diff
@@ -0,0 +1,290 @@
+		if (i >= t && j < t) return new BigDecimal(0.5);
+		i = i-t+1;
+		j = j-t+1;
+		BigDecimal half = new BigDecimal(0.5);
+		BigInteger sum = new BigInteger(0+"");
+		int max = (i > j)?i:j;
+		for (int k = 1; k <= 2*max+2; k++){
+			sum = sum.add(walks(1, k, i-1).multiply(walks(2, k, j-1)));
+			sum = sum.add(walks(2, k, i-1).multiply(walks(1, k, j-1)));
+		}
+		BigDecimal result = half.pow(i+j).multiply(new BigDecimal(sum.doubleValue()));
+		result = result.add(half);
+		return result;
+	}
+
+	private BigInteger walks(int i, int k, int j) {
+		if (j == 0){
+			if (i == k) return new BigInteger(1+"");
+			else return new BigInteger(0+"");
+		}
+		if (2*j < k - i) return new BigInteger(0+"");
+		if (k < i+j) return new BigInteger(0+"");
+		return Combinatory.comb(j, k-i-j);
+	}
+
+	
+	private BigDecimal computeProb2(int j, int i){
+		BigInteger den = (new BigInteger(""+2)).pow(i+j);
+		BigInteger t1, t2;
+		BigDecimal result = new BigDecimal(0+"");
+		int max = (i > j)?i:j;
+		for (int k = max; k <= 2*max; k++){
+			if (k-i > i) break;
+			if (k-j > j) break;
+			t1 = Combinatory.comb(i, k-i);
+			t2 = Combinatory.comb(j, k-j);
+			result = result.add(new BigDecimal((t1.multiply(t2))));
+		}
+		result = result.divide(new BigDecimal(den));
+		return result;
+	}
+
+	@Override
+	public DBProtocol getInstance() {
+		return new PoulidorProtocol();
+	}
+	
+	@Override
+	public BigDecimal getDistanceFraudProbability() {
+		/*if (n == 0) return BigDecimal.ONE;
+		if (n == 1) return new BigDecimal("0.75");
+		double p  = 1;
+		for (int i = 1; i <= n; i++){
+			p = p*(0.5 + computeProb2(i, i).doubleValue()/2);
+		}
+		BigDecimal half = new BigDecimal(0.5d);
+		BigDecimal four = new BigDecimal(4d);
+		BigDecimal sqrt = half.pow(2*n).add(((four).multiply(half.pow(n))).negate()).add(four.multiply(new BigDecimal(p)));
+		BigDecimal result = half.pow(n);
+		return new BigDecimal(result.doubleValue() + Math.sqrt(sqrt.doubleValue())/2);*/
+		//if (n == 0) return BigDecimal.ONE;
+		return new BigDecimal(distanceFraudFromDisk2(n));
+	}
+
+	@Override
+	public BigDecimal getTerroristFraudProbability() {
+		return ONE;
+	}
+
+	@Override
+	public boolean hasFinalSlowPhase() {
+		return false;
+	}
+
+	@Override
+	public int getCryptoCalls() {
+		return 1;
+	}
+
+	@Override
+	public DBProtocol[] getInstances() {
+		DBProtocol[] result = new DBProtocol[MAX_N];
+		for (int i = 0; i < MAX_N; i++){
+			result[i] = new PoulidorProtocol();
+			result[i].setNumberOfRounds(i+1);
+		}
+		return result;
+	}
+
+	@Override
+	public String getIdentifier() {
+		return "Poulidor-{"+n+"}";
+	}
+
+	@Override
+	public boolean lackSecurityProof() {
+		return true;
+	}
+
+	@Override
+	public int getSizeOfTheChannel() {
+		return 1;
+	}
+
+	
+	public static void generateDataFile() throws IOException{
+		String newLine = System.getProperty("line.separator");
+		PoulidorProtocol p = new PoulidorProtocol();
+		FileWriter writerMafia = new FileWriter(new File(p.getMafiaFileNameOfValues()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *writerMafia*. The resource is referenced at lines 174, 178. The resource is closed at line 181. There are other execution paths that do not contain closure statements, for example, when FileWriter constructor throws an exception. Ensure *writerMafia* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).